File: deploy_kr260/tests/test_dma_full.c
/*
 * test_dma_full.c - Test DMA completo del stream_adder (MM2S -> PL -> S2MM)
 *
 * Verifica el flujo completo:
 *   1. Aloca buffer contiguo via u-dma-buf (/dev/udmabuf0)
 *   2. Llena buffer de entrada con valores conocidos
 *   3. Programa AXI DMA MM2S para leer buffer y enviarlo al stream_adder
 *   4. Programa AXI DMA S2MM para recibir resultado y guardarlo en otra region
 *   5. Espera completado via polling de DMASR.IDLE
 *   6. Verifica que output = input + add_value
 *
 * Requiere:
 *   - xmutil loadapp stream_adder (bitstream cargado)
 *   - modulo u-dma-buf cargado: sudo insmod /lib/modules/.../u-dma-buf.ko udmabufs=udmabuf0=2097152
 *     (2 MB - suficiente para 512K samples de 32-bit)
 *   - sudo (para /dev/mem)
 *
 * Compilar: gcc -O2 -Wall -o test_dma_full test_dma_full.c
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>

#define DMA_BASE          0xA0000000UL
#define DMA_SIZE          0x10000UL
#define ADDER_BASE        0xA0010000UL
#define ADDER_SIZE        0x1000UL

/* Registros AXI DMA (PG021) */
#define REG_MM2S_DMACR    0x00
#define REG_MM2S_DMASR    0x04
#define REG_MM2S_SA       0x18  /* source addr */
#define REG_MM2S_LENGTH   0x28  /* transfer length */
#define REG_S2MM_DMACR    0x30
#define REG_S2MM_DMASR    0x34
#define REG_S2MM_DA       0x48  /* destination addr */
#define REG_S2MM_LENGTH   0x58

#define DMACR_RS          (1u << 0)
#define DMACR_RESET       (1u << 2)
#define DMASR_HALTED      (1u << 0)
#define DMASR_IDLE        (1u << 1)
#define DMASR_ERR_MASK    (0x70u << 4)  /* bits 4-6 */

#define NUM_SAMPLES       256
#define XFER_BYTES        (NUM_SAMPLES * 4)
#define BUFFER_SIZE       (2 * 1024 * 1024)  /* 2 MB total */
#define MM2S_OFFSET       0
#define S2MM_OFFSET       (1 * 1024 * 1024)  /* 2nd half */

#define ADD_VALUE         0x100  /* valor a sumar */

static volatile uint32_t *dma;
static volatile uint32_t *adder;
static uint32_t *udma_virt;
static uint64_t udma_phys;

static void dma_w(uint32_t o, uint32_t v) { dma[o/4] = v; __sync_synchronize(); }
static uint32_t dma_r(uint32_t o) { __sync_synchronize(); return dma[o/4]; }
static void adder_w(uint32_t o, uint32_t v) { adder[o/4] = v; __sync_synchronize(); }

static int read_u64(const char *path, uint64_t *out) {
    FILE *f = fopen(path, "r");
    if (!f) return -1;
    int n = fscanf(f, "%lx", (unsigned long *)out);
    fclose(f);
    return n == 1 ? 0 : -1;
}

static int wait_idle(uint32_t off, int timeout_ms) {
    while (timeout_ms-- > 0) {
        uint32_t sr = dma_r(off);
        if (sr & DMASR_ERR_MASK) {
            fprintf(stderr, "ERROR DMA SR=0x%08X\n", sr);
            return -1;
        }
        if (sr & DMASR_IDLE) return 0;
        usleep(1000);
    }
    return -1;
}

int main(void) {
    int fd_mem = open("/dev/mem", O_RDWR | O_SYNC);
    if (fd_mem < 0) { perror("/dev/mem"); return 1; }

    /* Mapear DMA y adder */
    void *p = mmap(NULL, DMA_SIZE, PROT_READ | PROT_WRITE,
                   MAP_SHARED, fd_mem, DMA_BASE);
    if (p == MAP_FAILED) { perror("mmap DMA"); return 1; }
    dma = (volatile uint32_t *)p;

    p = mmap(NULL, ADDER_SIZE, PROT_READ | PROT_WRITE,
             MAP_SHARED, fd_mem, ADDER_BASE);
    if (p == MAP_FAILED) { perror("mmap adder"); return 1; }
    adder = (volatile uint32_t *)p;

    /* Obtener direccion fisica del udmabuf */
    if (read_u64("/sys/class/u-dma-buf/udmabuf0/phys_addr", &udma_phys) < 0) {
        fprintf(stderr, "ERROR: no se puede leer phys_addr de udmabuf0\n");
        fprintf(stderr, "  -> Cargar el modulo: sudo modprobe u-dma-buf udmabufs=udmabuf0=%u\n",
                BUFFER_SIZE);
        return 1;
    }
    printf("udmabuf0 phys_addr = 0x%lX\n", udma_phys);

    /* Mapear el udmabuf */
    int fd_udma = open("/dev/udmabuf0", O_RDWR | O_SYNC);
    if (fd_udma < 0) { perror("/dev/udmabuf0"); return 1; }
    p = mmap(NULL, BUFFER_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd_udma, 0);
    if (p == MAP_FAILED) { perror("mmap udmabuf"); return 1; }
    udma_virt = (uint32_t *)p;
    printf("udmabuf0 mapeado: %p\n", udma_virt);

    /* Llenar buffer entrada */
    uint32_t *in_buf = (uint32_t *)((char *)udma_virt + MM2S_OFFSET);
    uint32_t *out_buf = (uint32_t *)((char *)udma_virt + S2MM_OFFSET);
    for (int i = 0; i < NUM_SAMPLES; i++) in_buf[i] = i;
    memset(out_buf, 0xFF, XFER_BYTES);   /* marcar como no escrito */

    /* Configurar stream_adder con ADD_VALUE */
    adder_w(0x00, ADD_VALUE);
    printf("stream_adder add_value = 0x%X\n", ADD_VALUE);

    /* Reset DMA */
    dma_w(REG_MM2S_DMACR, DMACR_RESET);
    while (dma_r(REG_MM2S_DMACR) & DMACR_RESET) usleep(10);
    dma_w(REG_S2MM_DMACR, DMACR_RESET);
    while (dma_r(REG_S2MM_DMACR) & DMACR_RESET) usleep(10);

    /* Arrancar canales */
    dma_w(REG_MM2S_DMACR, DMACR_RS);
    dma_w(REG_S2MM_DMACR, DMACR_RS);

    /* Programar S2MM primero (debe estar esperando antes de lanzar MM2S) */
    dma_w(REG_S2MM_DA, (uint32_t)(udma_phys + S2MM_OFFSET));
    dma_w(REG_S2MM_LENGTH, XFER_BYTES);

    /* Programar MM2S */
    dma_w(REG_MM2S_SA, (uint32_t)(udma_phys + MM2S_OFFSET));
    dma_w(REG_MM2S_LENGTH, XFER_BYTES);  /* escribir LENGTH lanza la transferencia */

    /* Esperar MM2S */
    if (wait_idle(REG_MM2S_DMASR, 5000) < 0) {
        fprintf(stderr, "TIMEOUT MM2S\n");
        return 1;
    }
    printf("MM2S completado\n");

    /* Esperar S2MM */
    if (wait_idle(REG_S2MM_DMASR, 5000) < 0) {
        fprintf(stderr, "TIMEOUT S2MM\n");
        return 1;
    }
    printf("S2MM completado\n");

    /* Verificar */
    int fail = 0;
    for (int i = 0; i < NUM_SAMPLES; i++) {
        uint32_t expected = i + ADD_VALUE;
        if (out_buf[i] != expected) {
            if (fail < 10)
                printf("[FAIL] out[%d] = 0x%08X (esperado 0x%08X)\n",
                       i, out_buf[i], expected);
            fail++;
        }
    }

    if (fail == 0) {
        printf("\n[OK] %d samples verificadas: out = in + 0x%X\n", NUM_SAMPLES, ADD_VALUE);
    } else {
        printf("\n[FAIL] %d/%d samples incorrectas\n", fail, NUM_SAMPLES);
    }

    munmap((void *)dma, DMA_SIZE);
    munmap((void *)adder, ADDER_SIZE);
    munmap(udma_virt, BUFFER_SIZE);
    close(fd_mem);
    close(fd_udma);

    return fail == 0 ? 0 : 1;
}

File: deploy_kr260/tests/test_dma_full.py
#!/usr/bin/env python3
"""
test_dma_full.py - Test DMA MM2S -> stream_adder -> S2MM en KR260

Usa /dev/mem (AXI-Lite control) + /dev/udmabuf0 (buffer contiguo CMA).

Requiere previo:
  1. xmutil loadapp stream_adder
  2. sudo modprobe u-dma-buf udmabufs=udmabuf0=2097152

Ejecutar:
  sudo python3 test_dma_full.py
"""

import mmap
import os
import struct
import sys
import time

DMA_BASE    = 0xA0000000
DMA_SIZE    = 0x10000
ADDER_BASE  = 0xA0010000
ADDER_SIZE  = 0x1000

# Registros DMA (PG021)
REG_MM2S_DMACR   = 0x00
REG_MM2S_DMASR   = 0x04
REG_MM2S_SA      = 0x18
REG_MM2S_LENGTH  = 0x28
REG_S2MM_DMACR   = 0x30
REG_S2MM_DMASR   = 0x34
REG_S2MM_DA      = 0x48
REG_S2MM_LENGTH  = 0x58

DMACR_RS    = 0x1
DMACR_RESET = 0x4
DMASR_IDLE  = 0x2
DMASR_ERR   = 0x70

NUM_SAMPLES = 256
XFER_BYTES  = NUM_SAMPLES * 4
BUFFER_SIZE = 2 * 1024 * 1024
MM2S_OFFSET = 0
S2MM_OFFSET = 1 * 1024 * 1024
ADD_VALUE   = 0x100


def rd(mem, off):
    mem.seek(off)
    return struct.unpack("<I", mem.read(4))[0]


def wr(mem, off, val):
    mem.seek(off)
    mem.write(struct.pack("<I", val))


def wait_idle(mem, off, timeout=5.0):
    t0 = time.time()
    while time.time() - t0 < timeout:
        sr = rd(mem, off)
        if sr & DMASR_ERR:
            raise RuntimeError(f"DMA error SR=0x{sr:08X}")
        if sr & DMASR_IDLE:
            return
        time.sleep(0.001)
    raise TimeoutError(f"DMA timeout, SR=0x{rd(mem, off):08X}")


def main():
    if os.geteuid() != 0:
        sys.exit("Requiere sudo")

    try:
        phys_addr = int(open("/sys/class/u-dma-buf/udmabuf0/phys_addr").read().strip(), 16)
    except FileNotFoundError:
        sys.exit("/dev/udmabuf0 no existe. Cargar: sudo modprobe u-dma-buf udmabufs=udmabuf0=2097152")

    print(f"udmabuf0 phys = 0x{phys_addr:X}")

    mf = open("/dev/mem", "r+b")
    dma = mmap.mmap(mf.fileno(), DMA_SIZE, offset=DMA_BASE)
    adder = mmap.mmap(mf.fileno(), ADDER_SIZE, offset=ADDER_BASE)

    uf = open("/dev/udmabuf0", "r+b")
    buf = mmap.mmap(uf.fileno(), BUFFER_SIZE)

    # Preparar datos
    for i in range(NUM_SAMPLES):
        buf.seek(MM2S_OFFSET + i*4)
        buf.write(struct.pack("<I", i))
    # Limpiar salida
    buf.seek(S2MM_OFFSET)
    buf.write(b"\xFF" * XFER_BYTES)

    # Configurar add_value
    wr(adder, 0x00, ADD_VALUE)
    print(f"add_value = 0x{ADD_VALUE:X}")

    # Reset DMA
    wr(dma, REG_MM2S_DMACR, DMACR_RESET)
    while rd(dma, REG_MM2S_DMACR) & DMACR_RESET: time.sleep(0.001)
    wr(dma, REG_S2MM_DMACR, DMACR_RESET)
    while rd(dma, REG_S2MM_DMACR) & DMACR_RESET: time.sleep(0.001)

    # Arrancar RS
    wr(dma, REG_MM2S_DMACR, DMACR_RS)
    wr(dma, REG_S2MM_DMACR, DMACR_RS)

    # Programar S2MM primero (receptor)
    wr(dma, REG_S2MM_DA, phys_addr + S2MM_OFFSET)
    wr(dma, REG_S2MM_LENGTH, XFER_BYTES)

    # Programar y lanzar MM2S
    wr(dma, REG_MM2S_SA, phys_addr + MM2S_OFFSET)
    wr(dma, REG_MM2S_LENGTH, XFER_BYTES)  # escritura de LENGTH = start

    print("Transferencia en curso...")
    wait_idle(dma, REG_MM2S_DMASR)
    print("MM2S OK")
    wait_idle(dma, REG_S2MM_DMASR)
    print("S2MM OK")

    # Verificar
    fail = 0
    for i in range(NUM_SAMPLES):
        buf.seek(S2MM_OFFSET + i*4)
        got = struct.unpack("<I", buf.read(4))[0]
        exp = (i + ADD_VALUE) & 0xFFFFFFFF
        if got != exp:
            if fail < 10:
                print(f"[FAIL] out[{i}] = 0x{got:08X} (esperado 0x{exp:08X})")
            fail += 1

    if fail == 0:
        print(f"\n[OK] {NUM_SAMPLES} samples correctas")
    else:
        print(f"\n[FAIL] {fail}/{NUM_SAMPLES}")

    buf.close(); uf.close()
    dma.close(); adder.close(); mf.close()
    sys.exit(0 if fail == 0 else 1)


if __name__ == "__main__":
    main()
